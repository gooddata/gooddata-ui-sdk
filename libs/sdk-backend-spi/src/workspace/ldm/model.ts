// (C) 2024 GoodData Corporation

/**
 * Interface that represents logical model date dataset.
 * @internal
 */
export interface IDateDataset {
    id: string;
    title: string;
    description?: string;
}

/**
 * Service for querying information about logical data model.
 *
 * This is internal service that returns data for internal UI application. It will be refactored once more
 * methods are added. This method would live in TigerSpecificFunctions, as the interface is experimental,
 * but Meditor has an issue with injecting both analytical backend and tiger specific functions objects.
 *
 * @internal
 */
export interface IWorkspaceLogicalModelService {
    /**
     * Returns information about date datasets in the workspace.
     * @param includeParents - true if parents should be included/
     */
    getDatasets(includeParents: boolean): Promise<IDateDataset[]>;
}
