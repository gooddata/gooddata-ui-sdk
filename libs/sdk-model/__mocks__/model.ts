/* tslint:disable:file-header */
/* tslint:disable:variable-name */
// THIS FILE WAS AUTO-GENERATED USING CATALOG EXPORTER; YOU SHOULD NOT EDIT THIS FILE; GENERATE TIME: 2019-10-17T14:42:41.833Z;
import {
    newAttribute,
    newMeasure,
    IAttribute,
    IMeasure,
    IMeasureDefinition,
    modifySimpleMeasure,
    newNegativeAttributeFilter,
} from "../src";

/*
 * Contents of this file were taken from the tools/reference-workspace project; sdk-model cannot depend on that
 * project since that would form a circular dependency between projects (ref workspace depends on sdk-model to get
 * the type definitions and factories and so on)
 *
 * Thus a subset of the model in the reference project is copied here directly.
 *
 * NOTE: If you find yourself in need for additional attributes / measures to be included in this testing model,
 * then copy more stuff from reference workspace. We DO NOT want any 'custom', 'dummy', 'mock' measures here.
 */

export const Account = {
    /**
     * Display Form Title: Name
     * Display Form ID: label.account.id.name
     */
    Name: newAttribute("label.account.id.name"),
    /**
     * Display Form Title: Account
     * Display Form ID: label.account.id
     */ Default: newAttribute("label.account.id"),
};
export const Activity = {
    /**
     * Display Form Title: Subject
     * Display Form ID: label.activity.id.subject
     */
    Subject: newAttribute("label.activity.id.subject"),
    /**
     * Display Form Title: Activity
     * Display Form ID: label.activity.id
     */ Default: newAttribute("label.activity.id"),
};
/**
 * Attribute Title: Activity Type
 * Display Form ID: attr.activity.activitytype
 */
export const ActivityType: IAttribute = newAttribute("label.activity.activitytype");
/**
 * Attribute Title: Department
 * Display Form ID: attr.owner.department
 */
export const Department: IAttribute = newAttribute("label.owner.department");

/**
 * Metric Title: Win Rate
 * Metric ID: aaX0PIUzg7nF
 * Metric Type: MAQL Metric
 */
export const WinRate: IMeasure<IMeasureDefinition> = newMeasure("aaX0PIUzg7nF");
/**
 * Metric Title: Won
 * Metric ID: afSEwRwdbMeQ
 * Metric Type: MAQL Metric
 */
export const Won: IMeasure<IMeasureDefinition> = newMeasure("afSEwRwdbMeQ");

/**
 * Fact Title: Duration
 * Fact ID: fact.stagehistory.duration
 */
export const Duration = {
    /**
     * Fact Title: Duration
     * Fact ID: fact.stagehistory.duration
     * Fact Aggregation: sum
     */
    Sum: newMeasure("fact.stagehistory.duration", (m) => m.aggregation("sum")),
    /**
     * Fact Title: Duration
     * Fact ID: fact.stagehistory.duration
     * Fact Aggregation: count
     */ Count: newMeasure("fact.stagehistory.duration", (m) => m.aggregation("count")),
    /**
     * Fact Title: Duration
     * Fact ID: fact.stagehistory.duration
     * Fact Aggregation: avg
     */ Avg: newMeasure("fact.stagehistory.duration", (m) => m.aggregation("avg")),
    /**
     * Fact Title: Duration
     * Fact ID: fact.stagehistory.duration
     * Fact Aggregation: min
     */ Min: newMeasure("fact.stagehistory.duration", (m) => m.aggregation("min")),
    /**
     * Fact Title: Duration
     * Fact ID: fact.stagehistory.duration
     * Fact Aggregation: max
     */ Max: newMeasure("fact.stagehistory.duration", (m) => m.aggregation("max")),
    /**
     * Fact Title: Duration
     * Fact ID: fact.stagehistory.duration
     * Fact Aggregation: median
     */ Median: newMeasure("fact.stagehistory.duration", (m) => m.aggregation("median")),
    /**
     * Fact Title: Duration
     * Fact ID: fact.stagehistory.duration
     * Fact Aggregation: runsum
     */ Runsum: newMeasure("fact.stagehistory.duration", (m) => m.aggregation("runsum")),
};
/**
 * Fact Title: Velocity
 * Fact ID: fact.stagehistory.velocity
 */
export const Velocity = {
    /**
     * Fact Title: Velocity
     * Fact ID: fact.stagehistory.velocity
     * Fact Aggregation: sum
     */
    Sum: newMeasure("fact.stagehistory.velocity", (m) => m.aggregation("sum")),
    /**
     * Fact Title: Velocity
     * Fact ID: fact.stagehistory.velocity
     * Fact Aggregation: count
     */ Count: newMeasure("fact.stagehistory.velocity", (m) => m.aggregation("count")),
    /**
     * Fact Title: Velocity
     * Fact ID: fact.stagehistory.velocity
     * Fact Aggregation: avg
     */ Avg: newMeasure("fact.stagehistory.velocity", (m) => m.aggregation("avg")),
    /**
     * Fact Title: Velocity
     * Fact ID: fact.stagehistory.velocity
     * Fact Aggregation: min
     */ Min: newMeasure("fact.stagehistory.velocity", (m) => m.aggregation("min")),
    /**
     * Fact Title: Velocity
     * Fact ID: fact.stagehistory.velocity
     * Fact Aggregation: max
     */ Max: newMeasure("fact.stagehistory.velocity", (m) => m.aggregation("max")),
    /**
     * Fact Title: Velocity
     * Fact ID: fact.stagehistory.velocity
     * Fact Aggregation: median
     */ Median: newMeasure("fact.stagehistory.velocity", (m) => m.aggregation("median")),
    /**
     * Fact Title: Velocity
     * Fact ID: fact.stagehistory.velocity
     * Fact Aggregation: runsum
     */ Runsum: newMeasure("fact.stagehistory.velocity", (m) => m.aggregation("runsum")),
};
export const ClosedDate = {
    /**
     * Display Form Title: mm/dd/yyyy (Closed)
     * Display Form ID: closed.date.mmddyyyy
     */
    MmDdYyyy: newAttribute("closed.date.mmddyyyy"),
    /**
     * Display Form Title: yyyy-mm-dd (Closed)
     * Display Form ID: closed.date.yyyymmdd
     */ YyyyMmDd: newAttribute("closed.date.yyyymmdd"),
    /**
     * Display Form Title: m/d/yy (no leading zeroes) (Closed)
     * Display Form ID: closed.date.mdyy
     */ MDYy: newAttribute("closed.date.mdyy"),
    /**
     * Display Form Title: Long (Mon, Jan 1, 2010) (Closed)
     * Display Form ID: closed.date.long
     */ Long: newAttribute("closed.date.long"),
    /**
     * Display Form Title: dd/mm/yyyy (Closed)
     * Display Form ID: closed.date.ddmmyyyy
     */ DdMmYyyy: newAttribute("closed.date.ddmmyyyy"),
    /**
     * Display Form Title: dd-mm-yyyy (Closed)
     * Display Form ID: closed.date.eddmmyyyy
     */ DdMmYyyy_1: newAttribute("closed.date.eddmmyyyy"),
};
export const SnapshotDate = {
    /**
     * Display Form Title: mm/dd/yyyy (Snapshot)
     * Display Form ID: snapshot.date.mmddyyyy
     */
    MmDdYyyy: newAttribute("snapshot.date.mmddyyyy"),
    /**
     * Display Form Title: yyyy-mm-dd (Snapshot)
     * Display Form ID: snapshot.date.yyyymmdd
     */ YyyyMmDd: newAttribute("snapshot.date.yyyymmdd"),
    /**
     * Display Form Title: m/d/yy (no leading zeroes) (Snapshot)
     * Display Form ID: snapshot.date.mdyy
     */ MDYy: newAttribute("snapshot.date.mdyy"),
    /**
     * Display Form Title: Long (Mon, Jan 1, 2010) (Snapshot)
     * Display Form ID: snapshot.date.long
     */ Long: newAttribute("snapshot.date.long"),
    /**
     * Display Form Title: dd/mm/yyyy (Snapshot)
     * Display Form ID: snapshot.date.ddmmyyyy
     */ DdMmYyyy: newAttribute("snapshot.date.ddmmyyyy"),
    /**
     * Display Form Title: dd-mm-yyyy (Snapshot)
     * Display Form ID: snapshot.date.eddmmyyyy
     */ DdMmYyyy_1: newAttribute("snapshot.date.eddmmyyyy"),
};

/**
 * An adhoc measure with aggregation "sum"
 */
export const AggregateAdhoc = Duration.Sum;

/**
 * An adhoc measure with empty filters
 */
export const EmptyFiltersAdhoc = modifySimpleMeasure(Won, (m) => m.noFilters().defaultLocalId());

/**
 * An adhoc measure with non-empty filters
 */
export const NonEmptyFiltersAdhoc = modifySimpleMeasure(Won, (m) =>
    m.filters(newNegativeAttributeFilter(Account.Name, ["value"])).defaultLocalId(),
);

/**
 * An adhoc measure with a computeRatio of false
 */
export const FalseComputeRatioAdhoc = modifySimpleMeasure(Won, (m) => m.noRatio().defaultLocalId());

/**
 * An adhoc measure with a computeRatio of true
 */
export const TrueComputeRatioAdhoc = modifySimpleMeasure(Won, (m) => m.ratio().defaultLocalId());
