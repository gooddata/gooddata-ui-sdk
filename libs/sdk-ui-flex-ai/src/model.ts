// (C) 2024 GoodData Corporation
import { v4 as uuidv4 } from "uuid";

/**
 * @alpha
 */
export type BaseMessage = {
    id: string;
    created: number;
};

/**
 * @alpha
 */
export type UserMessage = BaseMessage & {
    role: "user";
    content: string;
};

export const isUserMessage = (message: Message): message is UserMessage => message && message.role === "user";

export const makeUserMessage = (content: string): UserMessage => ({
    id: uuidv4(),
    role: "user",
    created: Date.now(),
    content,
});

/**
 * @alpha
 */
export type SystemMessage = BaseMessage & {
    role: "system";
    content: string;
};

export const isSystemMessage = (message: Message): message is SystemMessage =>
    message && message.role === "system";

export const makeSystemMessage = (content: string): SystemMessage => ({
    id: uuidv4(),
    role: "system",
    created: Date.now(),
    content,
});

/**
 * @alpha
 */
export type AssistantMessage = BaseMessage & {
    role: "assistant";
    content: string;
};

export const isAssistantMessage = (message: Message): message is AssistantMessage =>
    message && message.role === "assistant" && message.content !== null;

export const makeAssistantMessage = (content: string): AssistantMessage => ({
    id: uuidv4(),
    role: "assistant",
    created: Date.now(),
    content,
});

/**
 * All messages that can be stored in history.
 * @alpha
 */
export type Message = UserMessage | SystemMessage | AssistantMessage;

/**
 * Only messages that can be displayed in the chat, including in verbose mode.
 * @alpha
 */
export type VisibleMessage = UserMessage | SystemMessage | AssistantMessage;
