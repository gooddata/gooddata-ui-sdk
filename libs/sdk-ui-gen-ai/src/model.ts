// (C) 2024-2025 GoodData Corporation
import { v4 as uuidv4 } from "uuid";
import {
    GenAIChatRoutingUseCase,
    GenAIChatInteractionUserFeedback,
    IGenAIVisualization,
    ISemanticSearchResultItem,
} from "@gooddata/sdk-model";

/**
 * @public
 */
export type TextContents = {
    type: "text";
    text: string;
};

/**
 * @internal
 */
export const isTextContents = (contents: Contents): contents is TextContents => contents.type === "text";

/**
 * @internal
 */
export const makeTextContents = (text: string): TextContents => ({
    type: "text",
    text,
});

/**
 * @public
 */
export type RoutingContents = {
    type: "routing";
    text: string;
    useCase: GenAIChatRoutingUseCase;
};

/**
 * @internal
 */
export const isRoutingContents = (contents: Contents): contents is RoutingContents =>
    contents.type === "routing";

/**
 * @internal
 */
export const makeRoutingContents = (text: string, useCase: GenAIChatRoutingUseCase): RoutingContents => ({
    type: "routing",
    text,
    useCase,
});

/**
 * @public
 */
export type SearchContents = {
    type: "search";
    text: string;
    searchResults: ISemanticSearchResultItem[];
};

/**
 * @internal
 */
export const isSearchContents = (contents: Contents): contents is SearchContents =>
    contents.type === "search";

/**
 * @internal
 */
export const makeSearchContents = (
    text: string,
    searchResults: ISemanticSearchResultItem[] = [],
): SearchContents => ({
    type: "search",
    text,
    searchResults,
});

/**
 * @public
 */
export type VisualizationContents = {
    type: "visualization";
    text: string;
    createdVisualizations: IGenAIVisualization[];
};

/**
 * @internal
 */
export const isVisualizationContents = (contents: Contents): contents is VisualizationContents =>
    contents.type === "visualization";

/**
 * @internal
 */
export const makeVisualizationContents = (
    text: string,
    createdVisualizations: IGenAIVisualization[] = [],
): VisualizationContents => ({
    type: "visualization",
    text,
    createdVisualizations,
});

/**
 * @public
 */
export type ErrorContents = {
    type: "error";
    text: string;
};

/**
 * @internal
 */
export const isErrorContents = (contents: Contents): contents is ErrorContents => contents.type === "error";

/**
 * @internal
 */
export const makeErrorContents = (text: string): ErrorContents => ({
    type: "error",
    text,
});

/**
 * @public
 */
export type Contents =
    | TextContents
    | RoutingContents
    | SearchContents
    | VisualizationContents
    | ErrorContents;

/**
 * @public
 */
export type BaseMessage = {
    /**
     * Server-side ID for the message.
     */
    id?: string;
    /**
     * Local ID for the message. We need the id right away for optimistic rendering.
     */
    localId: string;
    /**
     * A timestamp of the message creation.
     */
    created: number;
    /**
     * Specifies if the message processing was cancelled.
     */
    cancelled: boolean;
    /**
     * Specifies if the message processing is complete.
     */
    complete: boolean;
    /**
     * The contents of the message.
     */
    content: Contents[];
};

/**
 * @public
 */
export type UserMessage = BaseMessage & {
    role: "user";
};

/**
 * @internal
 */
export const isUserMessage = (message?: Message): message is UserMessage => message?.role === "user";

/**
 * @internal
 */
export const makeUserMessage = (content: Contents[]): UserMessage => ({
    localId: uuidv4(),
    role: "user",
    created: Date.now(),
    cancelled: false,
    complete: true,
    content,
});

/**
 * @public
 */
export type AssistantMessage = BaseMessage & {
    role: "assistant";
    feedback: GenAIChatInteractionUserFeedback;
};

/**
 * @internal
 */
export const isAssistantMessage = (message?: Message): message is AssistantMessage =>
    message?.role === "assistant";

/**
 * Create a new Assistant message. By default, mark it as incomplete.
 * @internal
 */
export const makeAssistantMessage = (
    content: Contents[],
    complete = false,
    id?: string,
    feedback: GenAIChatInteractionUserFeedback = "NONE",
): AssistantMessage => ({
    id: id,
    localId: uuidv4(),
    role: "assistant",
    created: Date.now(),
    cancelled: false,
    feedback,
    complete,
    content,
});

/**
 * All messages that can be stored in history.
 * @public
 */
export type Message = UserMessage | AssistantMessage;
