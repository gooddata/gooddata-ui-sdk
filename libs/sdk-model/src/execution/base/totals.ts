// (C) 2019-2022 GoodData Corporation

import isEmpty from "lodash/isEmpty.js";
import { IMeasure, measureLocalId } from "../measure/index.js";
import { attributeLocalId, IAttribute } from "../attribute/index.js";
import { invariant } from "ts-invariant";
import { Identifier } from "../../objRef/index.js";

/**
 * Supported types of totals.
 *
 * @public
 */
export type TotalType = "sum" | "avg" | "max" | "min" | "med" | "nat";

/**
 * Describes type and granularity for calculation of Totals.
 *
 * @remarks
 * Total is calculated for particular measure and on some granularity - specified by an attribute
 * by which the measure is sliced by.
 *
 * @public
 */
export interface ITotal {
    /**
     * Type of total calculation.
     */
    type: TotalType;

    /**
     * Local identifier of measure for which to calculate total.
     */
    measureIdentifier: Identifier;

    /**
     * Local identifier of attribute - specifies granularity of the calculation.
     */
    attributeIdentifier: Identifier;

    /**
     * Specifies custom name for the calculated total. This will be included in result metadata.
     */
    alias?: string;
}

//
// Type guards
//

/**
 * Type-guard checking whether an object is a Total.
 *
 * @public
 */
export function isTotal(obj: unknown): obj is ITotal {
    return (
        !isEmpty(obj) &&
        (obj as ITotal).type !== undefined &&
        (obj as ITotal).measureIdentifier !== undefined &&
        (obj as ITotal).attributeIdentifier !== undefined
    );
}

//
//
//

/**
 * Creates a new total.
 *
 * @param type - type of total, one of the enumerated types
 * @param measureOrId - measure instance OR measure local identifier
 * @param attributeOrId - attribute instance OR attribute local identifier
 * @param alias - provide custom name (alias) for the total; this will be included in the computed results
 * @returns new total
 * @public
 */
export function newTotal(
    type: TotalType,
    measureOrId: IMeasure | Identifier,
    attributeOrId: IAttribute | Identifier,
    alias?: string,
): ITotal {
    invariant(type, "total type must be specified");
    invariant(measureOrId, "measure or measure local id must be specified");
    invariant(attributeOrId, "attribute or attribute local id must be specified");

    const measureIdentifier = measureLocalId(measureOrId);
    const attributeIdentifier = attributeLocalId(attributeOrId);
    const aliasProp = alias ? { alias } : {};

    return {
        type,
        measureIdentifier,
        attributeIdentifier,
        ...aliasProp,
    };
}

/**
 * Tests whether total instance represents a native total = a roll-up total.
 *
 * @public
 */
export function totalIsNative(total: ITotal): boolean {
    invariant(total, "total must be specified");

    return total.type === "nat";
}
