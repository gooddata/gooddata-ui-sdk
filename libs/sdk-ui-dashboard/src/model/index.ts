// (C) 2021-2024 GoodData Corporation

/*
 * The public API of the Dashboard model is exported from here.
 *
 * What is exported:
 *
 * -  hooks do dispatch commands / call selectors
 * -  all selectors & the typing of state with which they work
 * -  all events & their types. Note: event factories are not exported on purpose. outside code should not be
 *    creating events
 * -  all commands, their types & command factories
 */

export {
    DashboardContext,
    ObjectAvailabilityConfig,
    DashboardConfig,
    DashboardModelCustomizationFns,
    DashboardTransformFn,
    FiltersInfo,
    ResolvedDashboardConfig,
    ResolvedEntitlements,
    ResolvableFilter,
    ResolvedDateFilterValues,
    IResolvedAttributeFilterValues,
    IResolvedDateFilterValue,
    IResolvedFilterValues,
    WidgetsOverlayFn,
    IDashboardWidgetOverlay,
} from "./types/commonTypes.js";
export {
    ICustomWidget,
    ICustomWidgetDefinition,
    ICustomWidgetBase,
    newCustomWidget,
    newDashboardItem,
    newDashboardSection,
    isCustomWidgetDefinition,
    isCustomWidgetBase,
    isCustomWidget,
    ExtendedDashboardItem,
    extendedWidgetDebugStr,
    ExtendedDashboardWidget,
    DashboardItemDefinition,
    StashedDashboardItemsId,
    ExtendedDashboardLayoutSection,
    RelativeIndex,
    ExtendedDashboardItemType,
    ExtendedDashboardItemTypes,
} from "./types/layoutTypes.js";
export {
    FilterOp,
    FilterOpReplaceAll,
    FilterOpUnignoreAttributeFilter,
    FilterOpIgnoreAttributeFilter,
    FilterOpReplaceAttributeIgnores,
    FilterOpDisableDateFilter,
    FilterOpEnableDateFilter,
    FilterOpUnignoreDateFilter,
    FilterOpIgnoreDateFilter,
    FilterOperations,
    WidgetFilterOperation,
    WidgetHeader,
    WidgetDescription,
} from "./types/widgetTypes.js";

export {
    BrokenAlertType,
    IBrokenAlertFilterBasicInfo,
    BrokenAlertDateFilterInfo,
    BrokenAlertAttributeFilterInfo,
    isBrokenAlertDateFilterInfo,
    isBrokenAlertAttributeFilterInfo,
} from "./types/alertTypes.js";

export { ICsvExportConfig, IExportConfig, IXlsxExportConfig } from "./types/exportTypes.js";

export {
    IConnectingAttribute,
    IDashboardAttributeFilterDisplayForms,
    IDashboardAttributeFilterParentItem,
    IParentWithConnectingAttributes,
    IUseAttributeElements,
} from "./types/attributeFilterTypes.js";

export { IDashboardDependentDateFilter } from "./types/dateFilterTypes.js";

export { DRILL_TO_URL_PLACEHOLDER } from "./types/drillTypes.js";
export {
    DashboardAccessibilityLimitation,
    IInaccessibleDashboard,
} from "./types/inaccessibleDashboardTypes.js";

export * from "./react/index.js";
export * from "./commands/index.js";
export * from "./events/index.js";
export * from "./queries/index.js";
export * from "./store/index.js";

export { selectDateDatasetsForInsight } from "./queryServices/queryInsightDateDatasets.js";
export { selectInsightAttributesMeta } from "./queryServices/queryInsightAttributesMeta.js";
export { selectDateDatasetsForMeasure } from "./queryServices/queryMeasureDateDatasets.js";

export {
    DashboardEventHandler,
    DashboardEventHandlerFn,
    anyEventHandler,
    anyDashboardEventHandler,
    singleEventTypeHandler,
    commandStartedEventHandler,
    commandFailedEventHandler,
    DashboardEventEvalFn,
} from "./eventHandlers/eventHandler.js";
export { newDrillToSameDashboardHandler } from "./eventHandlers/drillToSameDashboardHandlerFactory.js";
export * from "./headlessDashboard/index.js";

export { isTemporaryIdentity } from "./utils/dashboardItemUtils.js";
export { existBlacklistHierarchyPredicate } from "./utils/attributeHierarchyUtils.js";
