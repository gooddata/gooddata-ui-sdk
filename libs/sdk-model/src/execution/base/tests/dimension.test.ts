// (C) 2019-2020 GoodData Corporation

import {
    dimensionSetTotals,
    dimensionsFindItem,
    dimensionTotals,
    IDimension,
    MeasureGroupIdentifier,
    newDimension,
    newTwoDimensional,
} from "../dimension.js";
import { ITotal } from "../totals.js";
import { Account, Activity } from "../../../../__mocks__/model.js";

const Total1: ITotal = {
    type: "avg",
    measureIdentifier: "measureId",
    attributeIdentifier: "localId1",
};
const Total2: ITotal = {
    type: "avg",
    measureIdentifier: "measureId",
    attributeIdentifier: "localId2",
};
const DimensionWithTotals = newDimension(["localId1", Total1]);
const DimensionWithoutTotals = newDimension(["localId1"]);

describe("newDimension", () => {
    const Scenarios: Array<[string, any, any]> = [
        ["create empty dimension if no input provided", undefined, undefined],
        [
            "create a valid dimension without totals when totals arg undef",
            ["localId1", "localId2"],
            undefined,
        ],
        ["create a valid dimension without totals when totals arg empty", ["localId1", "localId2"], []],
        [
            "create a valid dimension when using attribute objects and localIds",
            [Account.Name, "explicitLocalId1"],
            [],
        ],
        [
            "create a valid dimension when input arg is mix of identifiers and totals",
            ["localId1", Total1],
            [],
        ],
        ["create a valid dimension when totals are in second arg ", ["localId1"], [Total1]],
        ["create a valid dimension when totals are in both args ", ["localId1", Total1], [Total2]],
    ];

    it.each(Scenarios)("should %s", (_desc, itemsArg, totalsArg) => {
        expect(newDimension(itemsArg, totalsArg)).toMatchSnapshot();
    });
});

describe("newTwoDimensional", () => {
    const Scenarios: Array<[string, any, any]> = [
        ["create two empty dimensions", [], []],
        ["create first dimension empty", [], ["localId1"]],
        ["create second dimension empty", ["localId1"], []],
        ["create both dimensions with correct ids", ["localId1"], ["localId2"]],
        [
            "create both dimensions with correct ids when mixing attribute and localId",
            ["localId1", Account.Name],
            [Activity.Subject, "localId2"],
        ],
        ["create first dimension with totals", ["localId1", Total1], ["localId2"]],
        ["create second dimension with totals", ["localId1"], ["localId2", Total1]],
        ["create both dimension with totals", ["localId1", Total1], ["localId2", Total2]],
        ["create first dimension with measuregroup", [MeasureGroupIdentifier], []],
        ["create second dimension with measuregroup", [], [MeasureGroupIdentifier]],
    ];

    it.each(Scenarios)("should %s", (_desc, dim1, dim2) => {
        expect(newTwoDimensional(dim1, dim2)).toMatchSnapshot();
    });

    it("should throw when measure group in both dimensions", () => {
        expect(() => newTwoDimensional([MeasureGroupIdentifier], [MeasureGroupIdentifier])).toThrow();
    });
});

describe("dimensionSetTotals", () => {
    const Scenarios: Array<[string, IDimension, any]> = [
        ["replace non-empty totals", DimensionWithTotals, Total2],
        ["set totals ", DimensionWithoutTotals, Total1],
        ["clear totals", DimensionWithTotals, []],
        ["clear totals when no second arg", DimensionWithTotals, undefined],
    ];

    it.each(Scenarios)("should %s", (_desc, dimArg, totalsArg) => {
        expect(dimensionSetTotals(dimArg, totalsArg)).toMatchSnapshot();
    });
});

describe("dimensionTotals", () => {
    const Scenarios: Array<[string, any, ITotal[]]> = [
        ["return empty totals if not in dim", DimensionWithoutTotals, []],
        ["return totals from dim", DimensionWithTotals, [Total1]],
    ];

    it.each(Scenarios)("should %s", (_desc, input, expected) => {
        expect(dimensionTotals(input)).toEqual(expected);
    });

    const InvalidScenarios: Array<[string, any]> = [
        ["dim undefined", undefined],
        ["dim null", null],
    ];

    it.each(InvalidScenarios)("should thrown when %s", (_desc, input) => {
        expect(() => dimensionTotals(input)).toThrow();
    });
});

describe("dimensionsFindItem", () => {
    const TestDimensions = newTwoDimensional(["localId1", "localId2"], ["localId3", "localId4", "localId2"]);

    const Scenarios: Array<[string, IDimension[], string]> = [
        ["find one item in first dimension", TestDimensions, "localId1"],
        ["find one item in second dimension", TestDimensions, "localId4"],
        ["find one item in both dimensions", TestDimensions, "localId2"],
        ["find no item", TestDimensions, "localId0"],
    ];

    it.each(Scenarios)("should %s", (_desc, dims, id) => {
        expect(dimensionsFindItem(dims, id)).toMatchSnapshot();
    });
});
