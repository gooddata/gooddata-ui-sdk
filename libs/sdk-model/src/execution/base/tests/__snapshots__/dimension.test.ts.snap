// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dimensionSetTotals > should clear totals 1`] = `
{
  "itemIdentifiers": [
    "localId1",
  ],
}
`;

exports[`dimensionSetTotals > should clear totals when no second arg 1`] = `
{
  "itemIdentifiers": [
    "localId1",
  ],
}
`;

exports[`dimensionSetTotals > should replace non-empty totals 1`] = `
{
  "itemIdentifiers": [
    "localId1",
  ],
  "totals": {
    "attributeIdentifier": "localId2",
    "measureIdentifier": "measureId",
    "type": "avg",
  },
}
`;

exports[`dimensionSetTotals > should set totals  1`] = `
{
  "itemIdentifiers": [
    "localId1",
  ],
  "totals": {
    "attributeIdentifier": "localId1",
    "measureIdentifier": "measureId",
    "type": "avg",
  },
}
`;

exports[`dimensionsFindItem > should find no item 1`] = `[]`;

exports[`dimensionsFindItem > should find one item in both dimensions 1`] = `
[
  {
    "dim": {
      "itemIdentifiers": [
        "localId1",
        "localId2",
      ],
    },
    "dimIdx": 0,
    "itemIdx": 1,
  },
  {
    "dim": {
      "itemIdentifiers": [
        "localId3",
        "localId4",
        "localId2",
      ],
    },
    "dimIdx": 1,
    "itemIdx": 2,
  },
]
`;

exports[`dimensionsFindItem > should find one item in first dimension 1`] = `
[
  {
    "dim": {
      "itemIdentifiers": [
        "localId1",
        "localId2",
      ],
    },
    "dimIdx": 0,
    "itemIdx": 0,
  },
]
`;

exports[`dimensionsFindItem > should find one item in second dimension 1`] = `
[
  {
    "dim": {
      "itemIdentifiers": [
        "localId3",
        "localId4",
        "localId2",
      ],
    },
    "dimIdx": 1,
    "itemIdx": 1,
  },
]
`;

exports[`newDimension > should create a valid dimension when input arg is mix of identifiers and totals 1`] = `
{
  "itemIdentifiers": [
    "localId1",
  ],
  "totals": [
    {
      "attributeIdentifier": "localId1",
      "measureIdentifier": "measureId",
      "type": "avg",
    },
  ],
}
`;

exports[`newDimension > should create a valid dimension when totals are in both args  1`] = `
{
  "itemIdentifiers": [
    "localId1",
  ],
  "totals": [
    {
      "attributeIdentifier": "localId1",
      "measureIdentifier": "measureId",
      "type": "avg",
    },
    {
      "attributeIdentifier": "localId2",
      "measureIdentifier": "measureId",
      "type": "avg",
    },
  ],
}
`;

exports[`newDimension > should create a valid dimension when totals are in second arg  1`] = `
{
  "itemIdentifiers": [
    "localId1",
  ],
  "totals": [
    {
      "attributeIdentifier": "localId1",
      "measureIdentifier": "measureId",
      "type": "avg",
    },
  ],
}
`;

exports[`newDimension > should create a valid dimension when using attribute objects and localIds 1`] = `
{
  "itemIdentifiers": [
    "a_label.account.id.name",
    "explicitLocalId1",
  ],
}
`;

exports[`newDimension > should create a valid dimension without totals when totals arg empty 1`] = `
{
  "itemIdentifiers": [
    "localId1",
    "localId2",
  ],
}
`;

exports[`newDimension > should create a valid dimension without totals when totals arg undef 1`] = `
{
  "itemIdentifiers": [
    "localId1",
    "localId2",
  ],
}
`;

exports[`newDimension > should create empty dimension if no input provided 1`] = `
{
  "itemIdentifiers": [],
}
`;

exports[`newTwoDimensional > should create both dimension with totals 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
    ],
    "totals": [
      {
        "attributeIdentifier": "localId1",
        "measureIdentifier": "measureId",
        "type": "avg",
      },
    ],
  },
  {
    "itemIdentifiers": [
      "localId2",
    ],
    "totals": [
      {
        "attributeIdentifier": "localId2",
        "measureIdentifier": "measureId",
        "type": "avg",
      },
    ],
  },
]
`;

exports[`newTwoDimensional > should create both dimensions with correct ids 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
    ],
  },
  {
    "itemIdentifiers": [
      "localId2",
    ],
  },
]
`;

exports[`newTwoDimensional > should create both dimensions with correct ids when mixing attribute and localId 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
      "a_label.account.id.name",
    ],
  },
  {
    "itemIdentifiers": [
      "a_label.activity.id.subject",
      "localId2",
    ],
  },
]
`;

exports[`newTwoDimensional > should create first dimension empty 1`] = `
[
  {
    "itemIdentifiers": [],
  },
  {
    "itemIdentifiers": [
      "localId1",
    ],
  },
]
`;

exports[`newTwoDimensional > should create first dimension with measuregroup 1`] = `
[
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
  {
    "itemIdentifiers": [],
  },
]
`;

exports[`newTwoDimensional > should create first dimension with totals 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
    ],
    "totals": [
      {
        "attributeIdentifier": "localId1",
        "measureIdentifier": "measureId",
        "type": "avg",
      },
    ],
  },
  {
    "itemIdentifiers": [
      "localId2",
    ],
  },
]
`;

exports[`newTwoDimensional > should create second dimension empty 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
    ],
  },
  {
    "itemIdentifiers": [],
  },
]
`;

exports[`newTwoDimensional > should create second dimension with measuregroup 1`] = `
[
  {
    "itemIdentifiers": [],
  },
  {
    "itemIdentifiers": [
      "measureGroup",
    ],
  },
]
`;

exports[`newTwoDimensional > should create second dimension with totals 1`] = `
[
  {
    "itemIdentifiers": [
      "localId1",
    ],
  },
  {
    "itemIdentifiers": [
      "localId2",
    ],
    "totals": [
      {
        "attributeIdentifier": "localId1",
        "measureIdentifier": "measureId",
        "type": "avg",
      },
    ],
  },
]
`;

exports[`newTwoDimensional > should create two empty dimensions 1`] = `
[
  {
    "itemIdentifiers": [],
  },
  {
    "itemIdentifiers": [],
  },
]
`;
